File: day 11/quicksort.go
package main

import "fmt"

func QuicksortInt(numbers []int) []int {
	if len(numbers) <= 1 {
		return numbers
	}

	pivot := numbers[0]

	left := []int{}
	right := []int{}

	for _, value := range numbers[1:] {
		if value <= pivot {
			left = append(left, value)
		} else {
			right = append(right, value)
		}
	}

	left = QuicksortInt(left)
	right = QuicksortInt(right)

	return append(append(left, pivot), right...)

}

func main() {

	numbers := [6]int{5, 6, 2, 1, 3, 9}
	fmt.Println(QuicksortInt(numbers[:]))

	numbers2 := [6]int{33, 55, 22, 900, 2000, 1}
	fmt.Println(QuicksortInt(numbers2[:]))

}

File: day 12/quicksort.dart
List<int> quicksort(List<int> numbers) {
  if (numbers.length <= 1) return numbers;

  var pivot = numbers[0];

  List<int> left = [];
  List<int> right = [];

  for (var value in numbers.sublist(1)) {
    if (value <= pivot) {
      left.add(value);
    } else {
      right.add(value);
    }
  }

  left = quicksort(left);
  right = quicksort(right);

  List<int> result = [];
  result.addAll(left);
  result.add(pivot);
  result.addAll(right);

  return result;
}

main() {
  print(quicksort([5, 6, 2, 1, 3, 9]));
  // print(quicksort([33, 55, 22, 900, 2000, 1]));
}
